src/backends/policy_engine/souffle/datalog_lowering_visitor.h

```diff
@@ -0,0 +1,118 @@
+#ifndef SRC_BACKENDS_POLICY_ENGINE_SOUFFLE_DATALOG_LOWERING_VISITOR_H_
+#define SRC_BACKENDS_POLICY_ENGINE_SOUFFLE_DATALOG_LOWERING_VISITOR_H_
+
+#include "src/common/logging/logging.h"
+#include "src/ir/attributes/attribute.h"
+#include "src/ir/attributes/int_attribute.h"
+#include "src/ir/attributes/string_attribute.h"
+#include "src/ir/datalog/operation.h"
+#include "src/ir/ir_visitor.h"
+#include "src/ir/module.h"
+#include "src/ir/operator.h"
+#include "src/ir/ir_traversing_visitor.h"
+
+namespace raksha::backends::policy_engine::souffle {
+
+// A class containing the Datalog facts produced by the IR translator in
+// structured (ie, C++ objects, not strings) form.
+class RakshaDatalogFacts {
+ public:
+  void AddIsOperationFact(ir::datalog::IsOperationFact fact) {
+    is_operation_facts_.push_back(std::move(fact));
+  }
+
+  std::string ToDatalogString() const {
+    return absl::StrJoin(
+        is_operation_facts_, "\n",
+        [](std::string *out, const ir::datalog::IsOperationFact &arg) {
+          absl::StrAppend(out, arg.ToDatalogString());
+        });
+  }
+
+ private:
+  std::vector<ir::datalog::IsOperationFact> is_operation_facts_;
+};
+
+ class DatalogLoweringVisitor : public ir::IRTraversingVisitor<DatalogLoweringVisitor> {
+ public:
+  // TODO: Dedup this from `DecoderContext` in the SQL frontend.
+  static constexpr absl::string_view kDefaultOutputName = "out";
+  DatalogLoweringVisitor(ir::SsaNames &ssa_names) : ssa_names_(ssa_names) {}
+  virtual ~DatalogLoweringVisitor() {}
+
+  void PreVisit(const ir::Operation &operation) override {
+    // We currently don't have any owner information when outputting IR. We
+    // don't need it yet, really, but we do need to output something.
+    constexpr absl::string_view kUnknownPrincipal = "UNKNOWN";
```

Comment:
Define this to be class level constant? Use `"sql"` instead of `"UNKNOWN"` for now?

Reply:
Fixed.